### ISSUE TITLE
feat: Traversing tab/buffer history (navigate throught the stack of previously active tabs)

### ISSUE THREAD

**Issue by maricn** [tridactyl/tridactyl#4580]
### `:tab` back and forth
I find `<C-^>` (`<c-6>` if you prefer) very useful to jump to previously open buffer. As I often have many open tabs (50+), it would be quite useful to not only navigate back-and-forth between the two last active buffers, but to be able to traverse deeper down the MRU (most recently used) buffers.

Of course, this would mean bidirectional traversal, like `:tab <` and `:tab >` to go back and forward through MRU tabs.

### `:tab` MRU list
Come to think of it, another related feature that could be useful would be `:tab-mru` that would open the tab list dialog but sort the tabs by their most recent active-ity instead of their natural order in firefox.

Regarding the command itself, perhaps it would be somewhat intuitive if `:tab #` would open such MRU list, and automatically select the first item to preserve backward compatibility, but if you'd type `:tab #2` it would select the 2nd item on the MRU stack.
 
What do you think?

**Comment by bovine3dom** [comment]
Your `:tab-mru` already exists, sort of, with `:set tabsort mru`, but maybe there's a reason why you would want to be able to access both orders without changing a setting.

The first suggestion is a duplicate of #2378 - it's something I would really like but I'm not really sure how to do it as each time you navigate to a tab it then becomes the most recently used one, so we would need to keep a separate list and decide what interactions with a tab count as "using" it. We have a similar problem with `g;`.

**Comment by maricn** [comment]
### `:tab` MRU list
Thanks for sharing the `:set tabsort mru`, didn't know about it. The global config issue is not a biggie.
<img width="820" alt="image" src="https://user-images.githubusercontent.com/3995223/216365008-00bbac05-68db-4f28-b759-3f93158d9619.png">

I think it used to be default for a while probably as I seem to remember having the tab ids all over the place. Anyways, that is also an issue, since there's a break-of-flow when I do `<b>` and then instead of saying "2nd most recent", I need to look for the tab in the MRU list and locate its ID, and then type it in. So that's one extra downside of this method that I experience immediately.

### `:tab` back and forth
Actually I believe having the MRU idea in my mind polluted the initial feature request I was about to make. Instead of thinking of this path of traversal as MRU, it makes more sense to think of it as "undo", but instead of it being "change" history it would be a "navigation" history. Navigation, ofc, not in the context of the `window.location` of the current tab, but in the sense of the active tab in the firefox window.

This approach avoids the issue you mentioned with hijacking the MRU spot by navigating to it. Imagine having the (let's call it) window-tab-history as goes: 2-5-13-8-1-3 (where 2 was the active tab when the window was opened/restored, and 3 is the currently active tab). Navigating back through this window-tab-history would bring you to tab 1, navigating further back to tab 8. Opening a new tab (tab 14) would then append the new tab over the current one, changing the window-tab-history to: 2-5-13-8-14[-1-3] (see below for [-1-3] explanation).

In this approach, there would be another design decision to make, if whenever a new tab is opened all the "forward" window-tab-history should be discarded or just left in place. I would not discard it personally.

Is that understandable? It's a pretty specific request, so not sure if my whims should be entertained, but if it solves your old dilemma how to approach #2378, maybe it can be added to the roadmap. Cheers!

**Comment by bovine3dom** [comment]
Yeah, I think that `#n` as you suggested is a nice syntax for getting to the n-th MRU'd tab without having to press Tab n times.

You have described the problem nicely I think but really I or someone else needs to find the time to sit down and think about how to translate that into code. In your design the only way of changing the tab-visited history would be to make a new tab, is that right? I think that would be potentially counter-intuitive. In my mind as soon as I interactive with a page it becomes the most recently used tab.

**Comment by treapster** [comment]
1. Just in case, there is a switch in Firefox settings which makes ctrl+tab work in MRU order. I can't imagine navigating tabs without it. But you likely know about it.
2. I think sorting tabs in MRU order all the time is more intuitive and convenient than sorting by creation time.
3. I tried to implement MRU sorting and ID assignment for :b when MRU order is enabled, but encountered very nasty bugs when It jumped to a seemengly random tab, and failed to debug it enough to open MR. I can post the current patch or open WIP MR if you are interested.

**Comment by treapster** [comment]
I looked into it again and found the source of the issue, so i'm opening MR

**Comment by maricn** [comment]
> Yeah, I think that `#n` as you suggested is a nice syntax for getting to the n-th MRU'd tab without having to press Tab n times.
> 
> You have described the problem nicely I think but really I or someone else needs to find the time to sit down and think about how to translate that into code. In your design the only way of changing the tab-visited history would be to make a new tab, is that right? I think that would be potentially counter-intuitive. In my mind as soon as I interact with a page it becomes the most recently used tab.

Absolutely, I understand it's a niche feature request and that you have lots of issues, so I appreciate the attention here.

Regarding the update of the tab-visited history, your mention of interacting with a page is something I didn't incorporate in my suggestion, but definitely see as a crucial part of the feature. Your suggestion makes sense in the MRU approach, but if we would continue on the idea of "undo/redo"-like history, then on change/visit tab, the tab-visited history would append the tab after the current tab in the history.

From my previous example of history: 2-5-13-8-14[-1-3] (where "14" is the current tab and "1" & "3" are tabs we have visited but navigated away by "undo"-ing the tab-visit), accessing tab "5" would make it 2-5-13-8-14-5-[-1-3].

In other words, if this would be treated not just as a "sorting of :tab results" but as a separate functionality, it would make sense to just append the visited tab to the tab-visited history. In such implementation, tabs could appear in multiple places in the history, further resembling "undo/redo" for tab visits.

**Comment by bovine3dom** [comment]
Having thought about this a bit more, we could maybe get a decent "flip through MRU tabs" command working by putting any tabs accessed in the last ~10 seconds at the bottom of the list. It would be annoying for ctrl+6 and :tab so it would have to be a separate command
